packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -0,0 +1,591 @@
+///
+/// ### Example
+/// ```rust
+/// # // we only test on nightly, therefore report is unused (so is render)
+/// # #![cfg_attr(not(nightly), allow(dead_code, unused_variables, unused_imports))]
+/// use std::io::ErrorKind;
+///
+/// use error_stack::{fmt::Emit, Report};
+///
+///
+/// Report::install_debug_hook::<u64>(|val, ctx| {
+///     let mut acc = ctx.get::<u64>().copied().unwrap_or(0);
+///     acc += *val;
+///
+///     let mut div = ctx.get::<f32>().copied().unwrap_or(1.0);
+///     div /= *val as f32;
+///
+///     ctx.insert(acc);
+///
+///     Emit::next(format!("{val} (acc: {acc}, div: {div})"))
+/// });
+///
+/// let report = Report::new(std::io::Error::from(ErrorKind::InvalidInput))
+///     .attach(2u64)
+///     .attach(3u64);
+///
+/// # owo_colors::set_override(true);
+/// # fn render(value: String) -> String {
+/// #     let backtrace = regex::Regex::new(r"Backtrace No\. (\d+)\n(?:  .*\n)*  .*").unwrap();
+/// #     let backtrace_info = regex::Regex::new(r"backtrace with (\d+) frames \((\d+)\)").unwrap();
+/// #
+/// #     let value = backtrace.replace_all(&value, "Backtrace No. $1\n  [redacted]");
+/// #     let value = backtrace_info.replace_all(value.as_ref(), "backtrace with [n] frames ($2)");
+/// #
+/// #     ansi_to_html::convert_escaped(value.as_ref()).unwrap()
+/// # }
+/// #
+/// # #[cfg(nightly)]
+/// # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__hookcontext_storage.snap")].assert_eq(&render(format!("{report:?}")));
+/// #
+/// println!("{report:?}");
+/// ```
+///
+/// <pre>
+#[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__hookcontext_storage.snap"))]
+/// </pre>
+///
+/// [`Debug`]: core::fmt::Debug
+#[repr(transparent)]
+pub struct HookContext<'a, T> {
+    parent: &'a mut HookContextImpl,
+    _marker: PhantomData<T>,
+}
+
+impl<T> HookContext<'_, T> {
+    /// This snippet (which can include line breaks) will be appended to the
+    /// main message.
+    ///
+    /// This is useful for dense information like backtraces, or span traces.
+    ///
+    /// [`alternate()`]: Self::alternate
+    /// [`Debug`]: core::fmt::Debug
+    pub fn attach_snippet(&mut self, snippet: impl Into<String>) {
+        self.parent.snippets.push(snippet.into());
+    }
+}
+
+impl<'a, T> HookContext<'a, T> {
+    /// Cast the [`HookContext`] to a new type `U`.
+    ///
+    /// The storage of [`HookContext`] is partitioned, meaning that if `T` and `U` are different
+    /// types the values stored in [`HookContext<T>`] will be separated from values in
+    /// [`HookContext<U>`].
+    ///
+    /// Most user-facing should never need to use this function, as function hooks are only able to
+    /// get a mutable reference to [`HookContext`].
+    /// This is not the case for a fallback function, which receives the context as value,
+    /// allowing for "dynamic" recasting.
```

Comment:
This section is not true anymore. Maybe describe, that in most situations this is not needed at all, only when a type should share the storage with other types. (And maybe also that this could be useful when implementing a generic fallback).

Reply:
fixed in [bb1ee2d](https://github.com/hashintel/hash/pull/794/commits/bb1ee2dd646210bd6525f77e0c742611729b3a70)